core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -304,7 +304,27 @@ public void onFailure(Throwable t) {
                 observer.waitForNextChange(new ClusterStateObserver.Listener() {
                     @Override
                     public void onNewClusterState(ClusterState state) {
-                        threadPool.executor(executor).execute(AsyncReplicaAction.this);
+                        transportService.sendRequest(clusterService.localNode(), transportReplicaAction, request, new EmptyTransportResponseHandler(ThreadPool.Names.SAME) {
+
+                            @Override
+                            public void handleResponse(TransportResponse.Empty response) {
+                                try {
+                                    channel.sendResponse(response);
+                                } catch (IOException e) {
+                                    logger.debug("failed to send retry on replica response, action [{}], request [{}]", e, actionName, request);
```

Comment:
also, using a non-inner class means we can free the memory (i.e. cluster state copy) rather then have them accumulate with every retry